### PR TITLE
Add announce_port support

### PR DESCRIPTION
**Pull request by 0xThiebaut**
The `announce_port` setting permits to overwrite the port passed along to trackers as the `&port=` parameter. If left as the default, the listening port is used. This setting is only meant for very special cases where a seed's listening port differs from the effectively exposed port (e.g., through external NAT-PMP).

This PR would be complementary to the https://github.com/arvidn/libtorrent/pull/7726 `listen_on_proxy` setting.

### CODE REVIEW

**Comment by 0xThiebaut** [comment]
#### Example use-case
To secure outbound connections, a proxy server on the same subnet can be used to tunnel requests through a VPN (e.g., using [pufferffish/wireproxy](https://github.com/pufferffish/wireproxy)). Some VPNs support NAT-PMP (e.g., [ProtonVPN](https://protonvpn.com/support/port-forwarding-manual-setup/#linux)) which could technically allow the proxy server to perform dynamic port forwarding back.

Through the `announce_port` setting, a tracker can be informed on which NAT-PMP port has been made available through the proxy while the proxy has been configured to direct the port to the actual listening port.
Announce Port

File: src/http_tracker_connection.cpp [hunk 1]
```diff
@@ -143,7 +144,7 @@ namespace libtorrent {
 				, escape_string({tracker_req().pid.data(), 20}).c_str()
 				// the i2p tracker seems to verify that the port is not 0,
 				// even though it ignores it otherwise
-				, tracker_req().listen_port
+				, announce_port ? announce_port : tracker_req().listen_port
```
Comment (arvidn) [inline src/http_tracker_connection.cpp:147]: I think a better place to override the listen port is in `session_impl.cpp` where the tracker request object is set up and queued.

https://github.com/arvidn/libtorrent/blob/RC_2_0/src/session_impl.cpp#L1306-L1319

Then you don't need to change the HTTP and UDP tracker logic (I don't think at least)

File: src/tracker_manager.cpp [hunk 2]
```diff
@@ -278,8 +278,15 @@ constexpr tracker_request_flags_t tracker_request::i2p;
...
-			, req.listen_port);
+		if (cb) {
+			if (const auto announce_port = std::uint16_t(sett.get_int(settings_pack::announce_port))) {
+				cb->debug_log("*** QUEUE_TRACKER_REQUEST [ listen_port: %d, announce_port: %d ]"
+				, req.listen_port, announce_port);
+			} else {
+				cb->debug_log("*** QUEUE_TRACKER_REQUEST [ listen_port: %d ]"
+				, req.listen_port);
+			}
+		}
```
Comment (arvidn) [inline src/tracker_manager.cpp:289]: I'm not sure this change is necessary.

File: include/libtorrent/settings_pack.hpp [hunk 3]
```diff
@@ -2068,6 +2068,16 @@ namespace aux {
 			i2p_inbound_length,
 			i2p_outbound_length,
 
+			// ``announce_port`` is the port passed along to trackers as the
+			// ``&port=`` parameter. If left as the default, the listening port
+			// is used.
```
Comment (arvidn) [inline include/libtorrent/settings_pack.hpp:2073]: I think you need to mention that 0 means the listen port will not be overridden. I also think it would be nice to give an example of a setup where this would be useful.

**Review by arvidn — COMMENTED** [review]
I think it would be good to have some test coverage of this feature. Maybe you could add a test case here:

https://github.com/arvidn/libtorrent/blob/RC_2_0/simulation/test_tracker.cpp#L384-L410

File: src/tracker_manager.cpp (hunk 2 above)
Comment (0xThiebaut) [reply src/tracker_manager.cpp:289]: As in the entire change is unnecessary or you would rather have the a single `debug_log` call where `announce_port` would be `0` if not set?

Comment (arvidn) [reply src/tracker_manager.cpp:289]: I don't think it's important (nor necessary) to preserve the distinction between "listen_port" and "announce_port". We only use this for tracker announcements anyway.

Comment (0xThiebaut) [reply src/tracker_manager.cpp:289]: Thanks! The change will be dropped anyway as `req.listen_port` will be overwritten in https://github.com/arvidn/libtorrent/blob/RC_2_0/src/session_impl.cpp#L1306-L1319

Waiting on the tests to pass to submit the change.

**Comment by 0xThiebaut** [comment]
> I think it would be good to have some test coverage of this feature. Maybe you could add a test case here:
> 
> https://github.com/arvidn/libtorrent/blob/RC_2_0/simulation/test_tracker.cpp#L384-L410

A simple test coverage has been added to ensure `announce_ip` and `announce_port` are properly reflected in the tracker URL.

**Review by arvidn — COMMENTED** [review]
looks good. would you mind adding a line to `Changelog` too?

**Comment by 0xThiebaut** [comment]
> looks good. would you mind adding a line to `Changelog` too?

Done as well, thanks!

**Comment by arvidn** [comment]
should this feature affect the port we announce to the DHT?
the current default is to use "implied port", i.e. the source port of the DHT message. This works since we accept uTP connections on the same port.

**Comment by arvidn** [comment]
and I imagine it should *not* affect local service discovery announcements (i.e. multicast on the local network)

**Comment by 0xThiebaut** [comment]
The `announce_port` feature is intended for cases where additional appliances are located between local and remote peers. The local network discovery is by definition local so I agree it should not be affected by the `announce_port` setting. A future boolean setting could be introduced should anyone manage to find a use-case requiring `announce_port` overrides for local network discovery.

For DHT/uTP there can however be appliances between the local and remote peers; It would make sense to support overriding the announced port. I'm not familiar with DHT/uTP, do they currently use the same listening port as the TCP listener but for UDP?

**Comment by arvidn** [comment]
> I'm not familiar with DHT/uTP, do they currently use the same listening port as the TCP listener but for UDP?

Yes, libtorrent attempts to open the listen socket for TCP and UDP on the same port. The UDP socket is used for DHT, uTP (peer connections) and UDP trackers.

**Comment by 0xThiebaut** [comment]
My current understanding would be as follow:

For UDP trackers I think we are covered as it uses `req.listen_port` which `announce_port` overrides.
https://github.com/arvidn/libtorrent/blob/6a2c1eef6f3ba6b0a9763f3968bb79ddef73a355/src/udp_tracker_connection.cpp#L732

For uTP (peer connections) I don't believe there is a notion of "announcing"; it is merely the transport protocol and does not define which ports are used.

For DHT we need to ensure that `announce_port` indeed overrides the announced port. I believe we would need to do so in the session implementation and, if `announce_port` is set, make sure to clear the `dht::announce::implied_port` flag while overriding the `port` argument.
https://github.com/arvidn/libtorrent/blob/6a2c1eef6f3ba6b0a9763f3968bb79ddef73a355/src/session_impl.cpp#L6217-L6221

Would that make sense?

**Comment by 0xThiebaut** [comment]
Another DHT option is to do it at node-level where the logic already replaces `0` with the listening port; Although doing it at session-level would match the TCP `announce_port` variant.
https://github.com/arvidn/libtorrent/blob/6a2c1eef6f3ba6b0a9763f3968bb79ddef73a355/src/kademlia/node.cpp#L458-L467

**Comment by arvidn** [comment]
I can land this PR, and if you think the DHT behavior also should change, you could make another PR. I think having this logic in the session would make sense. It might also make sense to update the documentation for the setting to make it clear that it doesn't affect local service discovery and whether or not it affects DHT announces

**Comment by 0xThiebaut** [comment]
I'll update the PR today to reflect that `announce_port` does not affect local discovery. I'll also apply the DHT change as I would rather avoid complex documentation with non-intuitive edge-cases.

**Comment by 0xThiebaut** [comment]
I might be misunderstanding the relationship between a `session` and a `torrent` but wouldn't the following bypass the `announce_port` applied in the session's `session_impl::dht_announce`? Based on my limited understanding I would have expected the torrent to call `m_ses.dht_announce()` instead of `m_ses.dht()->announce()`.
https://github.com/arvidn/libtorrent/blob/e5fe95f92273bd775961e2f1f3b4ec45f8012d9e/src/torrent.cpp#L2816-L2820

**Comment by arvidn** [comment]
there's only one DHT node (technically, one per network interface you have). And to smooth-out the network traffic, torrents announces are staggered. The session loops over all the torrents and announces them one at a time, at a pace making them all announce once per 15 minutes (or whatever the setting has been set to).

**Comment by 0xThiebaut** [comment]
So on top of applying `announce_port` in `session_impl::dht_announce` we also need to add it to `torrent::dht_announce` as both functions determine a port to use? Or we move the logic to the underlying `dht_tracker::announce` which is used by both the `session_impl` and `torrent`?

Apologies if I am missing something.

File: src/session_impl.cpp [hunk 4]
```diff
@@ -6214,9 +6216,14 @@ namespace {
...
 
-	void session_impl::dht_announce(sha1_hash const& info_hash, int port, dht::announce_flags_t const flags)
+	void session_impl::dht_announce(sha1_hash const& info_hash, int port, dht::announce_flags_t flags)
 	{
 		if (!m_dht) return;
+		if (const auto announce_port = std::uint16_t(m_settings.get_int(settings_pack::announce_port)))
+		{
+			port = announce_port;
+			flags &= ~dht::announce::implied_port;
+		}
```
Comment (arvidn) [inline src/session_impl.cpp:6226]: is this change really necessary? you already have logic to not set `implied_port` and to pass in the overridden announce port, in `torrent.cpp` line 2008

File: ChangeLog [hunk 5]
```diff
@@ -13,6 +13,7 @@
 	* fix integer overflow in piece picker
 	* torrent_status::num_pieces counts pieces passed hash check, as documented
 	* check settings_pack::max_out_request_queue before performance alert
+	* add announce_port setting to overwrite the port announced to trackers
```
Comment (arvidn) [inline ChangeLog:16]:
```suggestion
	* add announce_port setting to override the port announced to trackers
```

File: include/libtorrent/settings_pack.hpp [hunk 6]
```diff
@@ -2068,6 +2068,20 @@ namespace aux {
 			i2p_inbound_length,
 			i2p_outbound_length,
 
+			// ``announce_port`` is the port passed along to trackers as the
+			// ``port`` parameter. This setting does not affect the effective
```
Comment (arvidn) [inline include/libtorrent/settings_pack.hpp:2072]: I think it would be good to mention that it affects DHT announcements as well

**Comment by arvidn** [comment]
thanks for your patience. I just did one more review pass

File: src/session_impl.cpp (hunk 4 above)
Comment (0xThiebaut) [reply src/session_impl.cpp:6226]: I believe so? Based on my understanding `dht_tracker::announce` is called directly from both `torrent::dht_announce` as well as `session_impl::dht_announce`. Hence, I added the `announce_port` in both of these places.

I don't believe there is a relationship between `torrent::dht_announce` and `session_impl::dht_announce` where one would affect the other?

Comment (arvidn) [reply src/session_impl.cpp:6226]: I see

Comment (arvidn) [reply src/session_impl.cpp:6226]: as far as I can tell, this function is only called from `session_handle.cpp`, which means the *user* is calling it. It might be a bit unintuitive to override the port and flag passed in directly from the client.

Comment (0xThiebaut) [reply src/session_impl.cpp:6226]: Alright so ideally we would have the `announce_port` setting only affect announcements which are not _client_-issued?

If so, would that mean I roll back this change and only keep `torrent::dht_announce` for DHT and `session_impl::queue_tracker_request` for HTTP/UDP?

In which case it might also be worth documenting the non-application of the `announce_port` on the `session_handle`.
https://github.com/arvidn/libtorrent/blob/7f691246aa58d756fbcdc2f3a03b1623055148ac/include/libtorrent/session_handle.hpp#L502-L517

Thanks for your patience!

Comment (arvidn) [reply src/session_impl.cpp:6226]: yeah, I think that sounds good

**Comment by arvidn** [comment]
thanks!
